Timetable and TimetableService

TeoV00 opened the pull request:
## Main features developed
### TrainTimetable entity and TimetableManager
For the moment you can create timetable as follow:
 ```
val partialTimetable: Either[ClockTimeErrors, PartialTimetable] = 
  PartialTimetable(train = AV1000Train, startStation = stationA, departureTime = h(9).m(0))
partialTimetable.map:
    _.stopsIn(stationB, waitTime = 5)(railAV_10)
    .transitIn(stationC)(railAV_10)
    .stopsIn(stationD, waitTime = 10)(railAV_10)
    .arrivesTo(stationF)(railAV_10)
 ```
### ClockTime
You can create new clock time with constructor:
```
ClockTime(h = -12, m = 30)
```
or using `h` and `m` (should be like a mini DSL, to fix):
```
h(10).m(45)
```

You can  add two ClockTime using `add` or `+` operator:
```
h(0).m(59) + h(0).m(59)  // => h(1).m(58)
h(23).m(59) + h(0).m(1)
```

### Other main feature
- Trains has lengthSize (metric unit)
- adopting `coordinate` refactoring

## Commits
- **feat: TrainTimetable entity (TDD)**
- **feat: new entity ClockTime (TDD), adapting TrainTimetable**
- **refactor: using contex ETA estimator in Timetable creation, some docs**
- **feat: added capacity to compare ClockTime**
- **docs: added some doc comments in Times**
- **feat: init train TimetableManager**
- **docs: added docs of TimetableManager**
- **feat: added remove method to TimetableManager TDD**
- **refactor: train, technology, timetable managers moved into managers package, renamed Train service**
- **refactor: changed alias type name, eta strategy name**
- **chore: preparation before rebasing from main**
- **refactor: rebase from main, some renamed imports**
- **refactor: use of non generic entities in Timetable service**
- **refactor: use of real types in timetable service**
- **test: fixed timetable entity test, back to use station**
- **docs: doc comments in Timetables**
- **feat: added lenght (metric unit) to train and to wagon UseType**
- **test: new mocked entities useful in tests, TimetableService, Timetable tests**
- **test: improved test, new test env object, utility methods**
- **test: TimetableService test invalid station name**

TeoV00 commented:
Not all scala doc has compact format. 
It is a TODO for next pr.

TeoV00 commented on src/main/scala/ulisse/entities/timetable/MockedEntities.scala:
that files it will be deleted when official `AppState` is ready.
This trait can be used as a view on AppState, let me know what do you think

Fede802 replied:
Yes, the final AppState should look like that trait

Fede802 commented on src/main/scala/ulisse/applications/managers/TimetableManagers.scala:
```diff
@@ -0,0 +1,127 @@
+      * @return
+      *   If al least one timetable is saved returns List of TrainTimetables, otherwise a Left of `TimetableNotFound`
+      */
+    def tablesOf(trainName: String): Either[TimetableNotFound, List[TrainTimetable]]
```
Is not legit to return simply an empty list if there is no timetible associated to a train?

Fede802 replied:
Okay if you search for a non existing train (find behaviour) make sense to return either or option

TeoV00 replied:
I can rename name method in `findTables` or `findTablesBy` to make it more clear

Fede802 commented on src/main/scala/ulisse/applications/managers/TimetableManagers.scala:
```diff
@@ -0,0 +1,127 @@
+      * @return
+      *   Returns initialized TimetableManager
+      */
+    def apply(timetables: List[TrainTimetable]): TimetableManager =
```
So the check to make an initial valid state is a to do? Or it's made somewhere else?

TeoV00 replied:
Checks of manager's initial state is a todo.
(Now you can initialize manager with overlapping tables or not existing train and so on, it's very bad I know 😅)

Fede802 commented on src/main/scala/ulisse/applications/ports/TimetablePorts.scala:
```diff
@@ -0,0 +1,57 @@
+    def createTimetable(
+        trainName: String,
+        departureTime: ClockTime,
+        stations: List[(StationId, Option[Int])]
```
What is Option[Int] used for?

TeoV00 replied:
`Option[Int]` is waiting time into a station.

I will improve that because is not clear what it is

Fede802 commented on src/main/scala/ulisse/entities/timetable/Timetables.scala:
```diff
@@ -0,0 +1,137 @@
+  private type WaitTime = Int
+
+  /** Rail info `length` and [[TypeRoute]] used to estimate `ScheduleTime` of station. */
+  case class RailInfo(length: Length, typeRoute: TypeRoute)
```
That could become part of route instead of external data structure?

TeoV00 replied:
I could 👍🏼,
@JBmanu What do you think?

I made it to group that info needed in time estimation calculation and not pass the entire route.

Fede802 commented on src/main/scala/ulisse/entities/timetable/Timetables.scala:
```diff
@@ -0,0 +1,137 @@
+        travelMinutes <- Some((railInfo.length / Math.min(train.maxSpeed, railInfo.typeRoute.technology.maxSpeed)) * 60)
+        offsetTime    <- lastTime
+        travelStartTime <- offsetTime.departure
+        arrivingTime    <- h(travelStartTime.h).m(travelStartTime.m + travelMinutes.toInt).toOption
```
yield directly "h(travelStartTime.h).m(travelStartTime.m + travelMinutes.toInt)"

Fede802 replied:
the yield append the last map to computation so maybe is logically better to have ".map(travelStartingTime => h(travelStartTime.h).m(travelStartTime.m + travelMinutes.toInt))" instead of ".map(arrivingTime => arrivingTime)"

Fede802 replied:
ok yes, the .toOption is needed regardless the positioning of that function, btw since yield do the last map of the computation could make sense to move it there instead of leave .map(arrivingTime -> arrivingTime). You'll lost arrivingTime placeholder but you already know the meaning of the result reading function name. So if u want evaluate it

TeoV00 replied:
if I put arrivingTime calculation directly in yield the returned type became `Option[Option[ClockTime]` and I need to unpack or flat in someway.

`h(travelStartTime.h).m(travelStartTime.m + travelMinutes.toInt)` returns an Either

Fede802 replied:
make sense

Fede802 commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,132 @@
+    *
+    * ClockTimes are compared by hours (`h`) and then by minutes (`m`) if hours are equals.
+    */
+  given Ordering[ClockTime] with
```
given Ordering[ClockTime] = Ordering.by(ct => (ct.h, ct.m)), maybe is possible to write something like this but it's to be checked

Fede802 commented on src/test/scala/ulisse/applications/managers/TimetableManagerTest.scala:
```diff
@@ -0,0 +1,106 @@
+class TimetableManagerTest extends AnyFeatureSpec with GivenWhenThen:
+
+  import ulisse.utils.Times.ClockTimeErrors
+  extension (timetable: Either[ClockTimeErrors, TrainTimetable])
```
Isn't this the "in" function?

TeoV00 replied:
Oh yes, I forgot that !! Thanks! I will fix this

Fede802 commented on src/main/scala/ulisse/entities/timetable/Timetables.scala:
```diff
@@ -0,0 +1,137 @@
+  private object UnrealTimeEstimator extends TimeEstimator:
+    def ETA(lastTime: Option[Time], railInfo: RailInfo, train: Train): Option[ClockTime] =
+      for
+        travelMinutes <- Some((railInfo.length / Math.min(train.maxSpeed, railInfo.typeRoute.technology.maxSpeed)) * 60)
```
idk if it sus to force monadic computation using the "Some" trick

TeoV00 replied:
its a trick, I removed it and created a val

JBmanu commented on src/main/scala/ulisse/entities/timetable/ScheduleTime.scala:
```diff
@@ -0,0 +1,36 @@
+  def departure: Option[ClockTime]
+
+object ScheduleTime:
+  def apply(arriving: Option[ClockTime], waitTime: Option[WaitTime], departure: Option[ClockTime]): ScheduleTime =
```
Why use optional values even in the implementation? 
You could put the checks in the apply method with the associated errors. That way the implementation would work only with the correct values without working with optionals each time.

TeoV00 replied:
**Option is not used to handle error, but is used because value should not be present.**

Because it is used to save departure, arrival and wait time for each stations (departing, arriving and those that are in the middle). 
The departing station (not the one of route but of entire timetable, for example in  A-> B -> C  A is departing station and C is the arriving one) has only `departure` time and other are None.

TeoV00 replied:
I changed name of entity into a better one. Is `TrainStationTime`

JBmanu commented on src/main/scala/ulisse/entities/timetable/Timetables.scala:
```diff
@@ -0,0 +1,137 @@
+    def table: ListMap[Station, Time]
+
+  /** Complete timetable containing arriving station and time */
+  trait TrainTimetable extends Timetable:
```
What is the motivation behind the splitting of the timetable concept? 
The timetable should always contain the train, the departure and arrival station, and the schedules, because they work together. This separation increases the complexity and usa

TeoV00 replied:
`PartialTimetable` represent a train timetable that has not yet completed/confirmed.
In few word its a timetable builder that when you call method `arrivesTo(station: Station)` you finalise table and get TrainTimetable (a complete timetable).

The relations between entities inside `Timetables` object are the following (look at image):
<img width="1158" alt="image" src="https://github.com/user-attachments/assets/8cdbaf7d-bc9d-4ac7-8ff7-8bd3308f021c" />

Type `Time` is `TrainTableTime`.

~~Maybe  `ScheduleTime` name is not so good and a better name should be `TableTime` or `StationTableTime` (Schedule word is better in simulation things )~~

TeoV00 replied:
TimeEstimator it is simply the strategy used for calculation of ETA (estimated time of arrival)

JBmanu commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,132 @@
+    override def toString: String = s"$h:$m"
+
+  object ClockTime:
+    def apply(h: Int, m: Int): Either[ClockTimeErrors, ClockTime] =
```
You can use a NonEmptyChain to manage and control errors

TeoV00 replied:
yess

JBmanu commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,132 @@
+  object ClockTime:
+    def apply(h: Int, m: Int): Either[ClockTimeErrors, ClockTime] =
+      for
+        h <- ValidationUtils.validateRange(h, 0, 23, InvalidHours())
```
Magic number

TeoV00 replied:
right

JBmanu commented on src/test/scala/ulisse/utils/ClockTimeTest.scala:
```diff
@@ -0,0 +1,36 @@
+      h(2).m(30) greaterThan h(10).m(45) should be(false)
+
+//
+//    "be subtracted to another ClockTimes" in :
```
What this annotated test means

TeoV00 replied:
its a test of un-implemented functionality.
I will remove comments and implement subtraction with clocktime

TeoV00 commented:
What in my last update:
- added checks on track availability in departure station (if at some time station is full, any other timetable can be saved)
- added checks in `TimetableManager` constructor (if some of timetables provided overlaps on another, the last computed is discarded)
-  `TrainTimetable` rename to `Timetable`

JBmanu commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,158 @@
+
+  object ClockTime:
+    private val maxDayHours        = 23
+    private val minDayHours        = 0
```
You can use range to min-max

TeoV00 replied:
To do what ? I just need to remove magic numbers so I created these vals. 

What do you mean `range to min-max`? I already use validation method that takes range

JBmanu replied:
Remember to do so, to be more explicit.

JBmanu commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,158 @@
+
+    private object FixedTimeDefault extends DefaultTimeStrategy:
+      override def defaultTime(currentTime: Time): Time = Time(0, 0, ignoredSecondValue)
+
```
Magic numbers

JBmanu commented on src/main/scala/ulisse/utils/Times.scala:
```diff
@@ -0,0 +1,158 @@
+        calculateSum(t, t2)
+
+    def greaterEqThan(time2: Either[ClockTimeErrors, ClockTime]): Boolean =
+      checkCondition(time, time2)(_ >= 0)
```
Magic number, equally for the next line.

TeoV00 replied:
Here 0 (zero) is not a time. Is just value used during comparison that can be -1, 0, 1 (is not my choice to use them to compare things)

JBmanu replied:
Why is it not your choice? however if you are referring to zero multiple times in the same context, you can create a variable so that you have neither magic numbers nor repeated code.

JBmanu commented on src/test/scala/ulisse/applications/useCases/TimetableServiceTest.scala:
```diff
@@ -0,0 +1,131 @@
+
+  Feature("User can create train timetables"):
+    Scenario("User creates new valid timetable"):
+      (TestEnvironment() and h(9).m(30)): (env, departureTime) =>
```
h(9).m(30) is repeated several times in the code, this also applies to h(8).m(30).

JBmanu commented on src/test/scala/ulisse/utils/TimesTest.scala:
```diff
@@ -0,0 +1,56 @@
+class TimesTest extends AnyWordSpec with Matchers:
+  "ClockTime" should:
+    "validate non in range values of hours and minutes" in:
+      ClockTime(h = -12, m = 30) should be(Left(InvalidHours(Time(-12, 30, 0))))
```
Why the syntax h(..).m(..) was not used.

TeoV00 replied:
You are asking me why? Because during developing I first created entity using constructor and I wanna keep in test to show that it can created also by constructor. Syntax `h().m()` is an extra thing.

JBmanu replied:
If they give the same behavior, that is perfectly fine.
